apps/rewards/app/components/Panel/NewReward/NewReward.js

```diff
@@ -641,6 +651,86 @@ class NewRewardClass extends React.Component {
     )
   }
 
+  getBlockProps = currentBlock => {
+    const {
+      amount,
+      amountToken,
+      rewardType,
+      dateReference,
+      dateStart,
+      dateEnd,
+      disbursement,
+      disbursementUnit,
+      disbursements,
+    } = this.state
+    const BLOCK_PADDING = 1
+    const amountBN = new BigNumber(amount)
+    const tenBN =  new BigNumber(10)
+    const decimalsBN = new BigNumber(amountToken.decimals)
+    const amountWei = amountBN.times(tenBN.pow(decimalsBN))
+    let startBlock = currentBlock + millisecondsToBlocks(Date.now(), dateStart)
+    let occurrences, isMerit, duration
+    if (rewardType === ONE_TIME_DIVIDEND || rewardType === ONE_TIME_MERIT) {
+      occurrences = 1
+    }
+    if (rewardType === ONE_TIME_MERIT) {
+      isMerit = true
+      duration = millisecondsToBlocks(dateStart, dateEnd)
+    } else {
+      isMerit = false
+    }
+    if (rewardType === RECURRING_DIVIDEND) {
+      occurrences = disbursements.length
+      switch (disbursementUnit) {
+      case 'Days':
+        duration = millisecondsToBlocks(Date.now(), disbursement * MILLISECONDS_IN_A_DAY + Date.now())
+        break
+      case 'Weeks':
+        duration = millisecondsToBlocks(Date.now(), disbursement * MILLISECONDS_IN_A_WEEK + Date.now())
+        break
+      case 'Years':
+        duration = millisecondsToBlocks(Date.now(), disbursement * MILLISECONDS_IN_A_YEAR + Date.now())
+        break
+      default:
+        duration = millisecondsToBlocks(Date.now(), disbursement * MILLISECONDS_IN_A_MONTH + Date.now())
+      }
+      startBlock -= duration
+    }
+    if(rewardType === ONE_TIME_DIVIDEND){
+      const rawBlockDuration = millisecondsToBlocks(Date.now(), dateReference)
+      startBlock = dateReference <= new Date() ? currentBlock + rawBlockDuration - BLOCK_PADDING : currentBlock 
+      duration = dateReference <= new Date() ? BLOCK_PADDING : rawBlockDuration
+    }
+    this.setState({
+      isMerit,
+      amountWei,
+      startBlock,
+      duration,
+      occurrences,
+    })
```

Comment:
I find it unexpected that `getBlockProps` has a side effect (setting state). Could we set this state in `setDisbursementBlocks`, with the other `setState`?

Reply:
Also, if we remove the call to `this.setState`, it becomes pretty easy to move this function outside of the component.